src/model/model.nit

```diff
@@ -1201,6 +1273,265 @@ abstract class MType
 	end
 end
 
+# A type that represents an operation over a set of types.
+abstract class MTypeSet[E: MType]
+	super MType
+
+	# The context in which this type was created.
+	#
+	# Used to populate `model` and look for `Object`.
+	#
+	# TODO: Remove this attribute once `as_notnull` don’t need it.
+	var mmodule: MModule
+
+	# The operands of this operation.
+	#
+	# Must not be empty. Must not be edited.
+	var operands: Set[E]
+
+	init do
+		assert operands.not_empty
+	end
+
+	redef fun model do return mmodule.model
+
+	redef fun location do return operands.first.location
+	# TODO: Return a more accurate location.
+
+	# Apply the same operator over `operands`.
+	#
+	# Does not consider properties of `self`. Used to abstract type
+	# generation done by methods of this class.
+	#
+	# May simplify the resulting expression and not return an instance of
+	# `SELF`. The contextual parameters (`mmodule` and `anchor`) are used to
+	# resolve types during comparisons in order to simplify the resulting type.
+	protected fun apply_to(operands: Set[MType], mmodule: MModule,
+			anchor: nullable MClassType): MType is abstract
```

Comment:
The problem I have with making this method private is that it indirectly makes `MTypeSet` “final” for all the other modules. Since `apply_to` is abstract, any subclass must redefine it. So, it makes sense to me to make this method visible to any subclass.